p2p/src/run_p2p.rs

```diff
@@ -0,0 +1,385 @@
+
+        self.peer_manager_mbox_tx
+            .send(PeerManagerEvent::Disconnect(peer_id, tx))
+            .map_err(|_| P2pError::ChannelClosed)?;
+        rx.await.map_err(P2pError::from)?
+    }
+
+    async fn get_peer_count(&self) -> crate::Result<usize> {
+        let (tx, rx) = oneshot_nofail::channel();
+        self.peer_manager_mbox_tx
+            .send(PeerManagerEvent::GetPeerCount(tx))
+            .map_err(P2pError::from)?;
+        rx.await.map_err(P2pError::from)
+    }
+
+    async fn get_bind_addresses(&self) -> crate::Result<Vec<String>> {
+        let (tx, rx) = oneshot_nofail::channel();
+        self.peer_manager_mbox_tx
+            .send(PeerManagerEvent::GetBindAddresses(tx))
+            .map_err(P2pError::from)?;
+        rx.await.map_err(P2pError::from)
+    }
+
+    async fn get_connected_peers(&self) -> crate::Result<Vec<ConnectedPeer>> {
+        let (tx, rx) = oneshot_nofail::channel();
+        self.peer_manager_mbox_tx
+            .send(PeerManagerEvent::GetConnectedPeers(tx))
+            .map_err(P2pError::from)?;
+        rx.await.map_err(P2pError::from)
+    }
+
+    async fn add_reserved_node(&mut self, addr: String) -> crate::Result<()> {
+        let addr = addr
+            .parse::<T::Address>()
+            .map_err(|_| P2pError::ConversionError(ConversionError::InvalidAddress(addr)))?;
+        self.peer_manager_mbox_tx
+            .send(PeerManagerEvent::AddReserved(addr))
+            .map_err(|_| P2pError::ChannelClosed)?;
+        Ok(())
+    }
+
+    async fn remove_reserved_node(&mut self, addr: String) -> crate::Result<()> {
+        let addr = addr
+            .parse::<T::Address>()
+            .map_err(|_| P2pError::ConversionError(ConversionError::InvalidAddress(addr)))?;
+        self.peer_manager_mbox_tx
+            .send(PeerManagerEvent::RemoveReserved(addr))
+            .map_err(|_| P2pError::ChannelClosed)?;
+        Ok(())
+    }
+
+    async fn submit_transaction(&mut self, tx: SignedTransaction) -> crate::Result<()> {
+        let id = tx.transaction().get_id();
+        self.mempool_handle.call_mut(|m| m.add_transaction(tx)).await??;
+        self.messaging_handle.broadcast_message(SyncMessage::NewTransaction(id))
+    }
+
+    fn subscribe_to_events(
+        &mut self,
+        handler: Arc<dyn Fn(P2pEvent) + Send + Sync>,
+    ) -> crate::Result<()> {
+        self.p2p_event_handlers_tx.send(handler).map_err(Into::into)
+    }
+}
+
+mod utils {
+
+    use std::{future::Future, time::Duration};
+
+    use futures::{future::BoxFuture, never::Never, FutureExt};
+    use logging::log;
+    use tokio::sync::oneshot;
+
+    use crate::Result;
+
+    type Killswitch = oneshot::Sender<Never>;
+
+    pub fn spawn_with_killswitch<F>(f: F) -> (BoxFuture<'static, Result<()>>, Killswitch)
+    where
+        F: Future<Output = Result<()>> + Send + 'static,
+    {
+        let (ks_tx, ks_rx) = killswitch_chan();
+        let jh = tokio::spawn(async move {
```

Comment:
I tried not spawning these activities, but instead "selecting" the collection of task-futures.

This did not work smoothly: the peer-manager and sync-manager do not get polled after the p2p-loop returns from the first `select_all`.

The networking-service backend is polled normally though.

Maybe somewhere in the peer-manager and sync-manager event-loops the waker call is "lost".

Reply:
Now that everything is spawned, it would probably make sense to use `JoinHandle::abort(&self)` for last-resort cancellation of a task :\